Sources/NIOConcurrencyHelpers/NIOLock.swift

```diff
@@ -0,0 +1,127 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+import Darwin
+#elseif os(Windows)
+import ucrt
+import WinSDK
+#else
+import Glibc
+#endif
+
+/// A threading lock based on `libpthread` instead of `libdispatch`.
+///
+/// This object provides a lock on top of a single `pthread_mutex_t`. This kind
+/// of lock is safe to use with `libpthread`-based threading models, such as the
+/// one used by NIO. On Windows, the lock is based on the substantially similar
+/// `SRWLOCK` type.
+public struct NIOLock {
+    @usableFromInline
+    internal let _storage: _Storage
+    
+    @usableFromInline
+    internal final class _Storage {
+        // TODO: We should tail-allocate the pthread_t/SRWLock.
+#if os(Windows)
+        @usableFromInline
+        internal let mutex: UnsafeMutablePointer<SRWLOCK> =
+        UnsafeMutablePointer.allocate(capacity: 1)
+#else
+        @usableFromInline
+        internal let mutex: UnsafeMutablePointer<pthread_mutex_t> =
+        UnsafeMutablePointer.allocate(capacity: 1)
+#endif
+        
+        /// Create a new lock.
+        public init() {
+#if os(Windows)
+            InitializeSRWLock(self.mutex)
+#else
+            var attr = pthread_mutexattr_t()
+            pthread_mutexattr_init(&attr)
+            debugOnly {
+                pthread_mutexattr_settype(&attr, .init(PTHREAD_MUTEX_ERRORCHECK))
+            }
+            
+            let err = pthread_mutex_init(self.mutex, &attr)
+            precondition(err == 0, "\(#function) failed in pthread_mutex with error \(err)")
+#endif
+        }
+        
+        deinit {
+#if os(Windows)
+            // SRWLOCK does not need to be free'd
+#else
+            let err = pthread_mutex_destroy(self.mutex)
+            precondition(err == 0, "\(#function) failed in pthread_mutex with error \(err)")
+#endif
+            mutex.deallocate()
+        }
+    }
+    
+    /// Create a new lock.
+    public init() {
+        self._storage = _Storage()
+    }
+
+    /// Acquire the lock.
+    ///
+    /// Whenever possible, consider using `withLock` instead of this method and
+    /// `unlock`, to simplify lock handling.
+    public func lock() {
+#if os(Windows)
+        AcquireSRWLockExclusive(self._storage.mutex)
+#else
+        let err = pthread_mutex_lock(self._storage.mutex)
+        precondition(err == 0, "\(#function) failed in pthread_mutex with error \(err)")
+#endif
```

Comment:
Can we delegate the methods in here to methods on `self._storage`? Swift plays a bit fast and loose with the guarantees around object lifetimes, but in principle we could accidentally `deinit` the lock here after loading it from `_storage`, whereas `self` is generally guaranteed to be alive.